File: leetcode/1474-delete-n-nodes-after-m-nodes-of-a-linked-list/solution.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def deleteNodes(self, head: Optional[ListNode], m: int, n: int) -> Optional[ListNode]:
        curr = head

        while curr:
            # Keep m nodes.
            for _ in range(m - 1):
                if not curr:
                    return head
                curr = curr.next
            if not curr:
                return head

            # Delete the next n nodes.
            to_delete = curr.next
            for _ in range(n):
                if not to_delete:
                    break
                to_delete = to_delete.next
            curr.next = to_delete
            curr = to_delete

        return head

File: leetcode/1474-delete-n-nodes-after-m-nodes-of-a-linked-list/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteNodes(ListNode head, int m, int n) {
        ListNode curr = head;

        while (curr != null) {
            for (int i = 0; i < m - 1 && curr != null; i++) {
                curr = curr.next;
            }
            if (curr == null) {
                return head;
            }

            ListNode toDelete = curr.next;
            for (int i = 0; i < n && toDelete != null; i++) {
                toDelete = toDelete.next;
            }
            curr.next = toDelete;
            curr = toDelete;
        }

        return head;
    }
}
